src/registers/debug.rs

```diff
@@ -0,0 +1,580 @@
+//! Functions to read and write debug registers.
+
+#![deny(unsafe_op_in_unsafe_fn)]
+
+use core::convert::TryFrom;
+use core::ops::Range;
+
+use crate::VirtAddr;
+
+use bit_field::BitField;
+use bitflags::bitflags;
+
+/// Debug Address Register
+///
+/// Holds the address of a hardware breakpoint.
+pub trait DebugAddressRegister {
+    /// Reads the current breakpoint address.
+    fn read() -> VirtAddr;
+
+    /// Writes the provided breakpoint address.
+    fn write(addr: VirtAddr);
+}
+
+macro_rules! debug_address_register_struct {
+    ($Dr:ident) => {
+        /// Debug Address Register
+        ///
+        /// Holds the address of a hardware breakpoint.
+        #[derive(Debug)]
+        pub struct $Dr;
+    };
+}
+
+debug_address_register_struct!(Dr0);
+debug_address_register_struct!(Dr1);
+debug_address_register_struct!(Dr2);
+debug_address_register_struct!(Dr3);
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(transparent)]
+/// A valid debug address register number.
+///
+/// Must be between 0 and 3 (inclusive).
+pub struct DebugAddressRegisterNumber(u8);
```

Comment:
Have we considered making this an enumeration? Alternatively we could also add constants corresponding to each number.